syntax = "proto3";

package farm_ng.nexus;

message ConfigureRequest {
  oneof kind {
    ConfigureRequestUpdate update = 1;
    ConfigureRequestList list = 2;
  }
}

message ConfigureRequestUpdate { repeated Parameter params = 1; }

message Parameter {
  string node = 1;
  string param = 2;
  ParameterValue value = 3;
}

message ParameterWithProperties {
  string node = 1;
  string param = 2;
  ParameterValue value = 3;
  bool is_mutable = 4;
}

message ParameterValue {
  oneof kind {
    bool bool = 1;
    int64 int64 = 2;
    uint64 usize = 3;
    double float64 = 4;
    string string = 5;
    VecFloat64 vec_float_64 = 6;
  }
}

message VecFloat64 { repeated double entries = 1; }

message ConfigureRequestList {}

message ConfigureReply {
  oneof kind {
    ConfigureReplyFailure failure = 1;
    ConfigureReplySuccess success = 2;
    ConfigureReplyList list = 3;
  }
}

message ConfigureReplyFailure { string message = 1; }

message ConfigureReplySuccess {}

message ConfigureReplyList { repeated ParameterWithProperties params = 1; }